controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,410 @@
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if pool.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Add finalizer if it does not exist
+		if controllerutil.AddFinalizer(pool, infrav1.OpenStackFloatingIPPoolFinalizer) {
+			return ctrl.Result{}, r.Client.Update(ctx, pool)
+		}
+	} else {
+		// Handle deletion
+		return ctrl.Result{}, r.reconcileDelete(ctx, scope, pool)
+	}
+
+	patchHelper, err := patch.NewHelper(pool, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, pool); err != nil {
+			if reterr == nil {
+				reterr = fmt.Errorf("error patching OpenStackFloatingIPPool %s/%s: %w", pool.Namespace, pool.Name, err)
+			}
+		}
+	}()
+
+	if err := r.reconcileFloatingIPNetwork(scope, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.setIPStatuses(ctx, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	claims := &ipamv1.IPAddressClaimList{}
+	if err := r.Client.List(context.Background(), claims, client.InNamespace(req.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, claim := range claims.Items {
+		claim := claim
+		log := log.WithValues("claim", claim.Name)
+		if !claim.ObjectMeta.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			ipAddress := &ipamv1.IPAddress{}
+			if err := r.Client.Get(ctx, client.ObjectKey{Name: claim.Name, Namespace: claim.Namespace}, ipAddress); err == nil {
+				// IPAddress already exists, another reconciler is working on it
+				log.Info("IPAddress already exists, another reconciler is working on it")
+				continue
+			}
+
+			ip, err := r.getIP(scope, pool)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			ipAddress = &ipamv1.IPAddress{
+				ObjectMeta: ctrl.ObjectMeta{
+					Name:      claim.Name,
+					Namespace: claim.Namespace,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: claim.APIVersion,
+							Kind:       claim.Kind,
+							Name:       claim.Name,
+							UID:        claim.UID,
+						},
+					},
+				},
+				Spec: ipamv1.IPAddressSpec{
+					ClaimRef: corev1.LocalObjectReference{
+						Name: claim.Name,
+					},
+					PoolRef: corev1.TypedLocalObjectReference{
+						APIGroup: pointer.String(infrav1.GroupVersion.Group),
+						Kind:     pool.Kind,
+						Name:     pool.Name,
+					},
+					Address: ip,
+					Prefix:  32,
+				},
+			}
+
+			// If the ReclaimPolicy is Delete, we add the finalizer to the IPAddress if they are not pre-allocated
+			// this is to ensure that the IPAddress is deleted from openstack when the claim is deleted
+			if pool.Spec.ReclaimPolicy == infrav1.ReclaimDelete && !contains(pool.Spec.PreAllocatedFloatingIPs, ip) {
+				controllerutil.AddFinalizer(ipAddress, infrav1.DeleteFloatingIPFinalizer)
+			}
```

Comment:
I'd prefer we always added a Finalizer. Lets put the cleanup logic explicitly in the delete flow of the controller rather than implicitly in the k8s GC.